pkg/trace/agent/agent.go

```diff
@@ -202,105 +208,185 @@ func (a *Agent) loop() {
+	m.Source.SpansReceived.Add(int64(len(pt.TraceChunk.Spans)))
+	return nil
+}
 
-	for i := 0; i < len(p.Chunks()); {
-		chunk := p.Chunk(i)
-		if len(chunk.Spans) == 0 {
-			log.Debugf("Skipping received empty trace")
-			p.RemoveChunk(i)
-			continue
-		}
+func (a *Agent) blocklist(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+	if !a.Blacklister.Allows(pt.Root) {
+		m.Source.TracesFiltered.Inc()
+		m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
+		return fmt.Errorf("Trace rejected by ignore resources rules. root: %v", pt.Root)
+	}
+	return nil
+}
 
-		tracen := int64(len(chunk.Spans))
-		ts.SpansReceived.Add(tracen)
-		err := normalizeTrace(p.Source, chunk.Spans)
-		if err != nil {
-			log.Debugf("Dropping invalid trace: %s", err)
-			ts.SpansDropped.Add(tracen)
-			p.RemoveChunk(i)
-			continue
-		}
+func (a *Agent) tagFilter(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+	if filteredByTags(pt.Root, a.conf.RequireTags, a.conf.RejectTags) {
+		m.Source.TracesFiltered.Inc()
+		m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
+		return fmt.Errorf("Trace rejected as it fails to meet tag requirements. root: %v", pt.Root)
+	}
+	return nil
+}
 
-		// Root span is used to carry some trace-level metadata, such as sampling rate and priority.
-		root := traceutil.GetRoot(chunk.Spans)
-		normalizeChunk(chunk, root)
-		if !a.Blacklister.Allows(root) {
-			log.Debugf("Trace rejected by ignore resources rules. root: %v", root)
-			ts.TracesFiltered.Inc()
-			ts.SpansFiltered.Add(tracen)
-			p.RemoveChunk(i)
-			continue
+func eachSpan(fs ...SpanTransformer) Transformer {
+	return func(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
+		for _, span := range pt.TraceChunk.Spans {
+			for _, f := range fs {
+				if err := f(m, pt, span); err != nil {
+					return err
+				}
+			}
 		}
+		return nil
+	}
+}
 
-		if filteredByTags(root, a.conf.RequireTags, a.conf.RejectTags) {
-			log.Debugf("Trace rejected as it fails to meet tag requirements. root: %v", root)
-			ts.TracesFiltered.Inc()
-			ts.SpansFiltered.Add(tracen)
-			p.RemoveChunk(i)
-			continue
+func (a *Agent) applyGlobalTags(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+	for k, v := range a.conf.GlobalTags {
+		if k == tagOrigin {
+			pt.TraceChunk.Origin = v
+		} else {
+			traceutil.SetMeta(s, k, v)
 		}
+	}
+	return nil
+}
 
-		// Extra sanitization steps of the trace.
-		for _, span := range chunk.Spans {
-			for k, v := range a.conf.GlobalTags {
-				if k == tagOrigin {
-					chunk.Origin = v
-				} else {
-					traceutil.SetMeta(span, k, v)
-				}
-			}
-			if a.ModifySpan != nil {
-				a.ModifySpan(span)
-			}
-			a.obfuscateSpan(span)
-			Truncate(span)
-			if p.ClientComputedTopLevel {
-				traceutil.UpdateTracerTopLevel(span)
-			}
-		}
-		a.Replacer.Replace(chunk.Spans)
+func (a *Agent) modifySpan(m *api.Metadata, pt *traceutil.ProcessedTrace, s *pb.Span) error {
+	if a.ModifySpan != nil {
```

Comment:
Usually the happy path is on the left, so on all these functions, I'd reverse the condition and return early in case there is nothing to do.

Reply:
I did this in several places where it makes sense, but not in a couple I thought it didn't.

For example, I kept:
```
func (a *Agent) blocklist(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
	if !a.Blacklister.Allows(pt.Root) {
		m.Source.TracesFiltered.Inc()
		m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
		return fmt.Errorf("Trace rejected by ignore resources rules. root: %v", pt.Root)
	}
	return nil
}
```

To me, this is clearer than
```
func (a *Agent) blocklist(m *api.Metadata, pt *traceutil.ProcessedTrace) error {
	if a.Blacklister.Allows(pt.Root) {
		return nil
	}
	m.Source.TracesFiltered.Inc()
	m.Source.SpansFiltered.Add(int64(len(pt.TraceChunk.Spans)))
	return fmt.Errorf("Trace rejected by ignore resources rules. root: %v", pt.Root)
}
```

In the second case, the "happy path" appears to return an error, which is odd, especially because the `return err` is disjointed from the condition that signals the error.

In the original, the happy path is `return nil`. Even though there's nothing to do, I think it still fits with "happy path on the left" since this *is* what we expect to happen most of the time. Also, to my ears, the human-language reading of "*if* the blacklister doesn't allow this *then* mark an error" is better than "*if* the blacklister allows it *then* return. Mark an error."

WDYT?